File: leetcode/31. 下一个排列.cpp
/*
 * @Author       : ApassEr
 * @Date         : 2020-06-15 16:58:05
 * @LastEditTime : 2020-06-15 17:08:58
 * @Description  : 31. 下一个排列
 * @Algorithm    : 参考STL:next_permutation源码
 */
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        if (nums.size() <= 1) return;
        auto before = nums.end() - 1;
        for (;;) {
            auto after = before--;
            if (*before < *after) {
                auto upper = nums.end();
                while (!(*before < *--upper));
                iter_swap(before, upper);
                reverse(after, nums.end());
                return;
            }
            if (before == nums.begin()) {
                reverse(nums.begin(), nums.end());
                return;
            }
        }
    }
};

File: leetcode/31. 下一个排列.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var nextPermutation = function(nums) {
    let len = nums.length;
    if (len <= 1) return;
    let before = len - 1;
    for (;;) {
        let after = before--;
        if (nums[before] < nums[after]) {
            let upper = len;
            while (!(nums[before] < nums[--upper]));
            [nums[before], nums[upper]] = [nums[upper], nums[before]];
            let part = nums.slice(after);
            part.reverse();
            nums.splice(after, len - after);
            for (let i of part) {
                nums.push(i);
            }
            return;
        }
        if (!before) {
            nums.reverse();
            return;
        }
    }
};